server-utils/src/cors.rs

```diff
@@ -1,9 +1,94 @@
 //! CORS handling utility functions
 
 use std::ascii::AsciiExt;
+use hosts::Host;
 
-// TODO [ToDr] Pre-process origins (should only contain protocol, host and port)
-pub type Origin = String;
+/// Origin Protocol
+#[derive(Clone, Hash, Debug, PartialEq, Eq)]
+pub enum OriginProtocol {
+	/// Http protocol
+	Http,
+	/// Https protocol
+	Https,
+	/// Custom protocol
+	Custom(String),
+}
+
+/// Request Origin
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Origin {
+	protocol: OriginProtocol,
+	host: Host,
+	as_string: String,
+}
+
+impl<T: AsRef<str>> From<T> for Origin {
+	fn from(string: T) -> Self {
+		Origin::parse(string.as_ref())
+	}
+}
+
+impl Origin {
+	fn with_host(protocol: OriginProtocol, host: Host) -> Self {
+		let string = Self::to_string(&protocol, &host);
+
+		Origin {
+			protocol: protocol,
+			host: host,
+			as_string: string,
+		}
+	}
+
+	/// Creates new origin given protocol, hostname and port parts.
+	/// Pre-processes input data if necessary.
+	pub fn new(protocol: OriginProtocol, host: &str, port: Option<u16>) -> Self {
+		Self::with_host(protocol, Host::new(host, port))
+	}
+
+	/// Attempts to parse given string as a `Origin`.
+	/// NOTE: This method always succeeds and falls back to sensible defaults.
+	pub fn parse(data: &str) -> Self {
+		let mut it = data.split("://");
+		let mut proto = it.next();
+		let mut hostname = it.next();
+
+		if hostname.is_none() {
+			hostname = proto;
+			proto = None;
+		}
+
+		let proto = proto.map(|s| s.to_lowercase());
+		let hostname = Host::parse(hostname.unwrap());
```

Comment:
it would be better to use `expect` here or even to refactor to:

```rust
let mut it = data.split("://").rev();
let mut hostname = Host::parse(it.next().expect("it.len() > 0; qed"));
let mut proto = it.next().map(String::to_lowercase);
```